### ISSUE TITLE
Setup of MatrixFree got slower since release 9.3

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#13411]
I ran a comparison of the run time of the new step_37 benchmark between the 9.2 release, the 9.3 release and current master. Here is a summary of results:
```
       grid  dofs  setmf rhs    transfer smooth mgsol1 mgsol2 mvd    mvf  mvfs
v9.2   0.378 0.519 0.314 0.0079 0.182    0.156  1.439  1.211  0.555 0.338 0.040
v9.3   0.378 0.523 0.315 0.0082 0.181    0.157  1.440  1.210  0.561 0.334 0.038
master 0.387 0.423 0.405 0.0079 0.184    0.148  1.404  1.177  0.557 0.342 0.040
```
Here, the first column is the setup of the grid, the second `distribute_dofs` and `distribute_mg_dofs`, whereas the third is the one of interest here, the setup of the `MatrixFree` objects for the active cells and the level cells. This went from 0.315 to 0.405 seconds and should thus get our attention (conversely, `mgsol1` and `mgsol2` which refer to solving, got a bit faster thanks to #12819, and might get an additional 10% boost if we implement renumbering of unknowns; the last three columns are matrix-vector products with double and float variables). The main reason appears to be the setup of the hanging node constraints implemented recently. Looking deeper, the function `HangingNodes::setup_lines_to_cell` is pretty expensive with many memory allocations. In step-37 with its uniform mesh, the information is not needed at all. Hence, a first solution would be to either disable the setup completely for uniformly refined meshes, or to only run it *near the edges* that actually are relevant. Now that is the main question, as we might not be able to deduce it from the neighbors alone when only a line is constrained. Nonetheless, I believe we should be able to do something. We might also be able to reduce the memory allocations by doing suitable `reserve` functions, but the more important step is to only compute it near the neighborhood where needed.

Another small thing is that https://github.com/dealii/dealii/blob/668fb9a63856716046da978e847aa4bafa7e550c/include/deal.II/matrix_free/hanging_nodes_internal.h#L754-L764
is called on every cell, whereas we should be able to do it in some setup function only once and store the supported components (and whether there is one supported component at all) as a class member.

**Comment by peterrum** [comment]
> Another small thing is that is called on every cell, whereas we should be able to do it in some setup function only once and store the supported components (and whether there is one supported component at all) as a class member.

PR https://github.com/dealii/dealii/pull/12617 is going in that direction

**Comment by kronbichler** [comment]
This issue has been addressed by a series of patches, namely #14311, #14115, #13951, #13884, #13873, #13872, #13869
